### DISCUSSION TITLE
Cannot finish alarm sound if using Bluetooth headset

### DISCUSSION THREAD

**Discussion by utzaki** [NightscoutFoundation/xDrip#1921] [Q&A]
If the forecast low alarm happens, while I am just using a Bluetooth headset, it is not possible to stop the alarm sound. 
All regular methods stopping the alarm sound do not work.
When switching off the headset the alarm continues on the mobile. Even hard stopping xDrip app does not stop the alarm. The only method that helps stopping the alarm sound is to restart the whole mobile. 
Pixel 4a, Android 12, xdrip 2021.11.06

Anyone the same experience or any idea?

**Comment by starbuck93** [comment]
I'm experiencing the same problem. I just sent feedback and logs through the app @jamorham. @Navid200 had a [similar Discussion in his old repository](https://github.com/Navid200/xDrip/discussions/110)  for this too. I just disabled the alarm sound in settings for predicted lows. I almost never have my ring volume up on my phone speakers, so it only happens with Bluetooth headbuds.

**Comment by Navid200** [comment]
@utzaki @starbuck93 There has been a recent update that may have addressed this on April 20, 2022.
Would you please try the latest Nightly and report if you still experience this issue?

**Reply by utzaki** [reply]
@Navid200 Installed last nightly and just encountered a low alarm. So I started bluetooth headset in time to check the issue. Indeed the behaviour changed. The alarm did not appear on the headset at all, so the problem did not came up. I will continue testing but until now it seems to be solved. Thanks so much! I will come back here in case there are still problems. Thanks. Thanks. Thanks.

**Reply by utzaki** [reply]
Oh, bad news. I just had the same problem. Low sugar alarm appeared while using the Bluetooth headset. It was only possible to stop by resetting the whole mobile. Sorry for that.

**Reply by utzaki** [reply]
I would love to be able to find the difference between yesterday and now. I am still using the same xDrip nightly. It was the same type of alarm. Only difference: Yesterday I used the headset only with my phone, now the headset was connected also to another audio source (Fire TV stick) that was active, while the mobile audio was passive. Hard to believe that this is relevant, but you never know.

**Reply by jamorham** [reply]
Which sound do you have set for the predicted low? Has this been customized or is it the system default? Either way can you try changing it to a sound which definitely doesn't repeat, eg a system notification sound which isn't intended as a ringtone and see if that makes any difference?

This class of alert simply attaches a sound to the system notification system. I don't really see why there would be any different handling when on bluetooth or otherwise. Also swiping off the notification should always cancel the sound as it is part of the notification.

Please also advise whether the notification channels feature has been enabled or not as there is slightly different handling internally for that.

**Reply by jamorham** [reply]
Please also clarify whether we are talking about predicted low or regular low alert we are talking about here as they use completely different audio techniques.

**Reply by utzaki** [reply]
It was currently the low alarm, not predicted low. 

I am using "xdrip default" as alarm tone. Here you will find all settings:
![Screenshot_20220522-192352](https://user-images.githubusercontent.com/53113714/169707651-651c4ced-d33d-40aa-ac91-fbc4810fa188.png)

Notification channels is activated.

**Reply by jamorham** [reply]
Which phone is this and which android version?

**Comment by utzaki** [comment]
Pixel 4a, Android 12

**Reply by jamorham** [reply]
Okay. I will attempt to reproduce this. Please confirm the following:

1) Settings as shown above for the alert
2) Low alarm triggers while bluetooth headphones are in use 
3) Alarms sounds on Phone speaker (not headphones)
4) Alarm refuses to stop

Does that correctly summarize the issue?

**Comment by utzaki** [comment]
1, 2, 4: Yes
3: I am not sure because I was using two audio sources on the headset and the headphone is always suppressing one of them. So maybe I only did not hear it on the headset, because it was supressed because the other audio source was just active. 

I will also try to reproduce it and would tell you if I may find out more.

**Reply by jamorham** [reply]
I've tested this with pixel 3a on Android 12 and been unable to reproduce the problem. I tested as follows using latest xDrip+ nightly.

1) Set up alarm as seen in screenshot
2) Connected bluetooth headphones
3) Started radio app which then plays via headphones
4) On the alert edit screen click test alert
5) Due to force speaker option this plays on the phone speaker, headphone pauses audio playback of radio
6) After alarm finished playing it stopped. Headphone radio resumed. 
7) In second test swiping off alert also silenced it.

There is one further option/code change that I can put in that we can try but ideally we want to be able to reproduce the problem. 

Does this problem happen every time? Does it happen when you use the test alert button?

**Reply by utzaki** [reply]
@jamorham "test alert" does not reproduce the problem here. That's the reason why it is a bit tricky to reproduce also for me.

@rodrigowebjump @starbuck93 If you still experience the problem please report also, if you are also using two audio sources simultaneously (multipoint pairing)  when the problem appears.

**Comment by Navid200** [comment]
@rodrigowebjump @starbuck93 
Do you still experience this problem?

**Reply by starbuck93** [reply]
The frequency that I had a predicted low alert and was wearing bluetooth earbuds was quite low, so I'll update xDrip now and see if I can get it to do it today.

**Comment by Navid200** [comment]
@utzaki There has been a change made to alert notification mechanism.
Would you please try the latest release (June 1, 2022) and report if you still experience the problem?
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.06.01

**Reply by utzaki** [reply]
@Navid200 I cannot test it before the end of June as I am on a trip now. I will do it as soon I am back.

**Reply by utzaki** [reply]
I am back and made a first test with the latest nightly. The alarm appeared and the behaviour was as it should be, so without endless alarm bug. 

To go sure I will make some more tests in the next days.

**Comment by thedewi** [comment]
I have been experiencing this a lot since I got one of the affected headsets around December, as covered in this archived discussion - https://github.com/Navid200/xDrip/discussions/135

Phone: Pixel 6 Pro
Affected headsets: Bose Noise Cancelling 700, Jabra Elite 7 Pro

This includes alarms like "forecast low" - which normally do not loop the ringtone - now loop without end despite dismissal.

This is very likely to be an audio playback channel issue, since it is associated with certain audio playback devices, and the sounds continue even if xDrip+ is force-stopped or paused.

xDrip+ gives every indication that the alarm has been dismissed, and does not mention any current alert. So the alarm *itself* is being dismissed okay, but the audio associated with the alert seems to be launching differently, so that it loops more, and continues without any involvement from the xDrip+ process.

**Reply by thedewi** [reply]
This is on current beta - e557cba-2022.05.19. I will try the nightly.

**Reply by Navid200** [reply]
Have you tried different MP3 files and experienced the same?  Or, have you only tried one?

**Reply by Navid200** [reply]
Can you try the latest Nightly released today?

**Reply by thedewi** [reply]
It's still occurring on the last two nightlies.

**Reply by thedewi** [reply]
I have now changed to a different sound, but there's really no reason to suspect an mp3 file could force an audio system to loop for certain specific output devices.

**Reply by Navid200** [reply]
If you have a sound device that keeps playing the file after the source, xDrip in this case, has stopped playing it, not much we can do in xDrip to change the behavior of your device.

**Reply by thedewi** [reply]
I agree it's clear that Android has some kind of bug, but it is only xDrip+ playing the sounds in a way that triggers this behaviour, and not hundreds of other apps, so there must be something specific about the way xDrip+ is playing sounds.

This has been going on for at least 9 months, and has had at least 3 threads about it, so it is happening on quite a few devices.

**Reply by jamorham** [reply]
Alarms like forecast low are simple notifications where we attach the sound id to the notification. In this situation xDrip has no control over sound playback after requesting the notification and I don't really see exactly how there could be an error there. The only thing I can think of is if somehow the ringtone etc being selected is marked as looping but if it only affects bluetooth headsets then this doesn't make sense either.

**Reply by thedewi** [reply]
I would like to attempt to investigate this on a code level. I have a lot of development experience, but not a lot in Android.

Can I create a debug build just by using Android Studio to do a gradle build?

Do you know what I should change to side-load this as a separate app - without affecting my existing xDrip+ installation? I could just change the application ID, right?

**Reply by mtomlinson** [reply]
I am also experiencing this issue, it does not occur with any of the more core alerts that I can snooze, it is only the alerts that are things like rising fast / falling fast and when I am listening to music while walking with my headphones. (pixel buds)

Happy to provide more details if any of this is helpful.

**Reply by michaelcadilhac** [reply]
Same here; it _sometimes_ occurs for hard alerts too, but I can snooze them using the Snooze Alert dialog.  This is with the latest APK.  This is very frustrating :(

**Reply by Navid200** [reply]
@michaelcadilhac Please explain what you mean by "hard alerts".

When this happens, what do you see in the xDrip logs?  Is there only one alert or is there a sequence of repeating alerts?

**Reply by michaelcadilhac** [reply]
I meant the low/high alerts.  Peaking at the logs, it may be intended; it reads:

[red] "Attempt to cancel alert (bg_alerts) within minimum limit of 2 seconds"

There's only one alert.

**Reply by Navid200** [reply]
Please tell us what phone and which Android version.

**Reply by michaelcadilhac** [reply]
- Pixel 5
- xDrip+ e13a76e-2022.08.03
- Android 13 TP1A.221105.002
- Sony WH-1000XM4

**Comment by thedewi** [comment]


**Reply by thedewi** [reply]
It's clearly an Android bug, since even force-stopping and disabling xDrip+ doesn't make the sound stop. I just can't understand how such an annoying problem has persisted so long.

**Reply by thedewi** [reply]
(But this is the only app I've ever experienced this problem with, so there must be some combination of flags in some API call that triggers the bug)

**Comment by utzaki** [comment]
Ever since I upgraded to Android 14 (Pixel 7a), I haven't encountered the bug again. While it might still be coincidental, it's worth noting that I frequently use earphones, which leads me to believe that the issue may have been resolved in Android 14.

**Reply by thedewi** [reply]
This sounds promising. I've been holding off upgrading to 14 just in case Diabox breaks, but I've heard a few reports it works okay, although none on my exact model (Pixel 6 Pro). Perhaps it's time for me to take the plunge and see.

**Comment by Navid200** [comment]
Does this ever happen without a headphone involved?

I now use a Pixel 6a.  But, I don't use a headphone.  I have never experienced this.

**Reply by thedewi** [reply]
For me it started happening as soon as I got a specific new Bluetooth headset (Bose Noise Cancelling 700), but has continued on other new models too (Jabra Elite 7 Pro), and I can't recall it ever happening without a Bluetooth headset paired. So I suspect it's some weird Android bug involving newer BT profiles and something specific about the way xDrip+ plays sounds. I still haven't found a chance to try getting a custom build going to randomly change some API call parameters.

**Reply by utzaki** [reply]
For me it is (was?) clearly linked with using a bluetooth headphone. I experienced the problem quite often and with two different bluetooth headphones but never without.

**Comment by jimbaxter** [comment]
This has started happening to me on an Android Pixel 8 with Bluetooth headset in the last few weeks.
Only a restart fixes it.

**Reply by Navid200** [reply]
I need to get a Bluetooth headset so that I can recreate this problem.
Can you point to a not very expensive option online so that I can see what I am looking for?

**Reply by Navid200** [reply]
OK, I got headphones.  I tested with a Motorola and the forecast low alert can be snoozed with no trouble.
But, I believe this problem occurs with Pixel.  

Tomorrow, I will test with a Pixel.

**Reply by Navid200** [reply]
I cannot recreate the problem with Pixel 6a.   
I wonder if it may have to do with the length of the sound file.  My sound file is extremely short.  It is just a few seconds.  I'm sure it is less than 5 seconds.  

How long (seconds) does your sound file last?

---  
  
If any of you likes to help, but don't know how to trigger a forecast low alert, please note that you can do the following:
1- Use a test phone for the test.  Don't perform the tests with a phone that you use with an active sensor.  
2- On the test phone, use fake data source: https://navid200.github.io/xDrip/docs/FakeDataSource.html
3- Speed up the fake data source and you will get a forecast low alert about once an hour.
4- Disable engineering mode after you are done with testing.

**Reply by Navid200** [reply]
By the way, I am running Android 15 beta on my Pixel 6a.

**Reply by michaelcadilhac** [reply]
This has happened to me once since moving to a Pixel 7a / Android 14, while it was happening much more often when I was on Pixel 5.  I use the [_Little Red Trolley_](https://www.zedge.net/ringtone/588ec8b7-8b96-4f53-b33e-e30a1a1a0269) sound as notification, which is fairly short.  I have not been able to reproduce this bug with any kind of consistency, sadly.

**Reply by Navid200** [reply]
I want to fix this if I know how.
I have alerts that use sound files as short of that if not shorter.

**Comment by xitation** [comment]
I've got same issue, it doesn't appear to matter if on Bluetooth headphones or not, and it's random when it happens. 

I'm on a Google Pixel 8 Pro happy to provide any other into needed to help, I'm pretty technical so if you need anything specific let me know.

**Comment by Navid200** [comment]
I wonder if we can put all the information together to see what is common between those who experience this problem.

I bring to your attention this old issue:
https://github.com/NightscoutFoundation/xDrip/issues/1859  

Eventually, it was discovered that the issue was being caused by a particular app.  
For all we know, something like that could be the cause of this as well.
But, until that was confirmed, there was no way to figure out it was being caused by a particular app.  And we are lucky that a user was able to identify the app.

If everyone who is experiencing this problem has the same app on their phone, we can ask them to uninstall the app to see if the problem goes away.

Assuming the problem is caused by an app, another way to deal with it is for one brave person to create a backup and restore their phone to factory settings.
Then, only install xDrip.  Does the problem persist?  If we are lucky and the answer is no, the brave soul can install one app a day until the problem returns.  This will let us identify the app.

I have no way of being sure this is caused by an app.
But, because some people are experiencing it and some, like me, are not, I wonder if that could caused by an app that some use and some don't.

**Reply by xitation** [reply]
It's odd though, it only appears to happen with the "predictive low alerts" for me. Not any of the others. They all snooze/stop correctly all the time.

**Reply by xitation** [reply]
Which somewhat points back to the logic of how that bit of the app works, maybe?

**Reply by Navid200** [reply]
Good point.  We can all look and see if we can identify what forecast low does that the rest of the alerts don't.

**Reply by xitation** [reply]
Is there additional logging we can use? 

Or if required can we build a debug version that captures that stuff we need that a few of us can run for a while until hopefully we catch the conditions that trigger it?

**Reply by Navid200** [reply]
There are additional logs you can enable.  That would bring in lots of logs.  But, these are logs that the developers have included.
Alternatively, we can add custom logs where we suspect may provide useful clues.

The challenge is that the problem persists even when you kill xDrip.  This means that even if there is a trigger in xDrip, the problem is visible outside xDrip.  We cannot add logs to xDrip to study what happens outside xDrip.

**Reply by xitation** [reply]
Hmm good point.

**Reply by xitation** [reply]
There is a notification log we can start to look at and try to corelate things against.

long-pressing an empty space on your home screen and then selecting Widgets. Scroll down to the Settings shortcut widget, long-press it and drag it onto your home screen. You'll then be prompted to select what the shortcut should target. Scroll down and select Notification log. 

Found this here - https://support.google.com/pixelphone/thread/3296640/repeating-unknown-notification-sound?hl=en

**Reply by xitation** [reply]
Keep in mind it looks like the log is flushed on reboot, so suggest taking a look at it / screenshot of the time window when things go wrong before rebooting to fix.

**Reply by utzaki** [reply]
Since June 2022, the problem has not occurred here again. Before that, it happened very frequently. Therefore, I'm afraid I can no longer help to identify the problem or possibly an app that causes it.
